src/include/gcc.h

```diff
@@ -27,71 +27,145 @@
- * swap) operations.
+ * A spin lock is an exclusive lock that generally expects all reads and writes on shared memory
+ * locations under its protection to be performed under the lock. In WiredTiger this pattern
+ * occasionally is not upheld and we read variables written to within the lock without locking. Such
+ * instances are only permitted if reading inconsistent data is sufficient for the use case.
+ *
+ * Read Write Locks:
+ * A read-write lock allows concurrent shared access for read-only operations, whereas write
+ * operations require exclusive access. This means that multiple threads can read the data in
+ * parallel but an exclusive lock is needed for writing or modifying data. When a writer is writing
+ * data, all other writers and readers will be blocked until the writer is finished writing.
+ * WiredTiger has its own read-write lock implementation. It is based on atomic operations and
+ * condition variables.
+ *
+ * It is expected that all the reads and writes of a shared variable protected by a read-write lock
+ * will be performed under the lock. For the same reason as spinlocks, WiredTiger may allow usages
+ * that read the shared variables without taking the lock. Such instances are only permitted if
+ * reading inconsistent data is sufficient for the use case.
+ *
+ * # Atomic operations  #
+ * As previously mentioned WiredTiger already expects loads and stores to specific sizes of points
+ * to be atomic and prevent tearing. WiredTiger also relies on atomic RMW operations, these include:
+ * - Compare and Swap
+ * - Fetch add / sub
+ * - Add / sub fetch
+ *
+ * Currently WiredTiger is undergoing a transition where all shared memory variable access will be
+ * wrapped by an atomic function, this will enable TSan to run on the code base.
+ *
+ * # Memory barriers #
+ * WiredTiger utilizes memory barriers in a number of places to ensure that CPU instruction
+ * reordering doesn't result in a concurrency bug. Specifically CPUs are allowed to reorder load and
+ * store instructions. If this occurs when two threads are reading or writing to the same shared
+ * memory locations it can result in those threads seeing what would otherwise be impossible values.
+ *
+ * WiredTiger defines barriers by describing which kind of reordering they prevent, for example a
+ * LoadLoad barrier prevents load instructions before the barrier from being reordered with load
+ * instructions after the barrier. Using this definition we can define 4 types of reorderings,
+ * LoadLoad, StoreStore, LoadStore and StoreLoad. To simplify the barrier semantics further
+ * WiredTiger defines acquire and release barriers, firstly we have the WT_ACQUIRE_READ_WITH_BARRIER
+ * macro which prevents LoadLoad and LoadStore reorderings. Next we have the
+ * WT_RELEASE_WRITE_WITH_BARRIER macro which prevents StoreStore and LoadStore reorderings, these
+ * barriers should be used as pairs. If one location requires an acquire barrier then there must be
```

Comment:
```suggestion
 * WT_RELEASE_WRITE_WITH_BARRIER macro which prevents StoreStore and LoadStore reorderings. These macros are always associated with a read or write to memory respectively. WT_ACQUIRE_READ_WITH_BARRIER ensures that all loads and stores after its associated read must take place after the read, and WT_RELEASE_WRITE_WITH_BARRIER enures that all loads and stores before its associated write take place before the write.
 * These barriers should be used as pairs. If one location requires an acquire barrier then there must be
```
This sentence is a bit messy, feel free to clean up as needed.